File: clients/dart/fiducia.dart
// Fiducia HTTP client (Dart). Zero-dependency — dart:io + dart:convert.
// Implements PROTOCOL.md.
//
//   final c = FiduciaClient("https://api.fiducia.cloud");
//   final lock = await c.lockAcquire("orders/checkout", ttlMs: 30000);
//   await c.lockRelease("orders/checkout", "worker-a", lock["result"]["output"]["fencing_token"]);

import 'dart:async';
import 'dart:convert';
import 'dart:io';

typedef FiduciaSyncJsonSender = Future<Map<String, Object?>> Function(
    Map<String, Object?> write);
typedef FiduciaSyncJsonPull = Future<Map<String, Object?>> Function(
    int cursor, int limit);

class FiduciaException implements Exception {
  final int status;
  final dynamic body;
  FiduciaException(this.status, this.body);
  @override
  String toString() => 'fiducia: HTTP $status';
}

class FiduciaRequestOptions {
  final Duration? timeout;
  final Duration? requestTimeout;
  final Duration? lockRequestTimeout;
  final int maxRetries;
  final int retryMax;
  final int retries;
  final Duration retryDelay;
  final String? idempotencyKey;

  const FiduciaRequestOptions({
    this.timeout,
    this.requestTimeout,
    this.lockRequestTimeout,
    this.maxRetries = 0,
    this.retryMax = 0,
    this.retries = 0,
    this.retryDelay = Duration.zero,
    this.idempotencyKey,
  });
}

class FiduciaClient {
  final String base;
  final HttpClient _http = HttpClient();
  Duration? requestTimeout;
  Duration? lockRequestTimeout;
  int retryMax = 0;
  Duration retryDelay = Duration.zero;

  FiduciaClient(String baseUrl) : base = baseUrl.replaceAll(RegExp(r'/+$'), '');

  Future<dynamic> _request(
    String method,
    String path, [
    Object? body,
    FiduciaRequestOptions? options,
    bool lockAcquire = false,
  ]) async {
    final maxRetries = _resolveRetries(options);
    for (var attempt = 0;; attempt++) {
      try {
        final timeout = _resolveTimeout(options, lockAcquire);
        final future = _requestOnce(method, path, body, options);
        return timeout == null ? await future : await future.timeout(timeout);
      } catch (e) {
        if (attempt >= maxRetries || !_retryable(e)) rethrow;
        final delay = _resolveRetryDelay(options);
        if (delay > Duration.zero) await Future<void>.delayed(delay);
      }
    }
  }

  Future<dynamic> _requestOnce(
    String method,
    String path, [
    Object? body,
    FiduciaRequestOptions? options,
  ]) async {
    final req = await _http.openUrl(method, Uri.parse(base + path));
    final idempotencyKey = options?.idempotencyKey;
    if (idempotencyKey != null) {
      if (idempotencyKey.trim().isEmpty ||
          idempotencyKey.contains(RegExp(r'[\r\n]'))) {
        throw ArgumentError.value(
          idempotencyKey,
          'idempotencyKey',
          'must be nonempty and contain no line breaks',
        );
      }
      req.headers.set('idempotency-key', idempotencyKey);
    }
    if (body != null) {
      req.headers.contentType = ContentType.json;
      req.write(jsonEncode(body));
    }
    final res = await req.close();
    final text = await res.transform(utf8.decoder).join();
    final data = text.isNotEmpty ? jsonDecode(text) : null;
    if (res.statusCode >= 300) throw FiduciaException(res.statusCode, data);
    return data;
  }

  Duration? _resolveTimeout(FiduciaRequestOptions? options, bool lockAcquire) =>
      options?.lockRequestTimeout ??
      options?.requestTimeout ??
      options?.timeout ??
      (lockAcquire ? lockRequestTimeout : null) ??
      requestTimeout;

  int _resolveRetries(FiduciaRequestOptions? options) {
    for (final value in [
      options?.maxRetries ?? 0,
      options?.retryMax ?? 0,
      options?.retries ?? 0,
      retryMax,
    ]) {
      if (value > 0) return value;
    }
    return 0;
  }

  Duration _resolveRetryDelay(FiduciaRequestOptions? options) =>
      options != null && options.retryDelay > Duration.zero
          ? options.retryDelay
          : retryDelay;

  bool _retryable(Object e) {
    if (e is FiduciaException) {
      return const [408, 425, 429, 500, 502, 503, 504].contains(e.status);
    }
    return e is IOException || e is TimeoutException;
  }

  String _enc(String s) => Uri.encodeComponent(s);

  FiduciaRequestOptions _withIdempotencyKey(
    FiduciaRequestOptions? options,
    String key,
  ) =>
      FiduciaRequestOptions(
        timeout: options?.timeout,
        requestTimeout: options?.requestTimeout,
        lockRequestTimeout: options?.lockRequestTimeout,
        maxRetries: options?.maxRetries ?? 0,
        retryMax: options?.retryMax ?? 0,
        retries: options?.retries ?? 0,
        retryDelay: options?.retryDelay ?? Duration.zero,
        idempotencyKey: key,
      );

  // --- misc ---
  Future<dynamic> health() => _request('GET', '/healthz');
  Future<dynamic> status() => _request('GET', '/v1/status');
  void close() => _http.close(force: true);

  // --- local-first sync ---
  Future<Map<String, Object?>> syncWrite(
    Map<String, Object?> write, {
    String pathPrefix = '/api/customer/sync',
    FiduciaRequestOptions? options,
  }) async {
    final table = write['table'];
    final id = write['id'];
    final operation = write['op'];
    final baseVersion = write['base_version'];
    if (table is! String || table.trim().isEmpty) {
      throw ArgumentError.value(table, 'write.table', 'must be nonempty');
    }
    if (id is! String || id.trim().isEmpty) {
      throw ArgumentError.value(id, 'write.id', 'must be nonempty');
    }
    if (operation != 'upsert' && operation != 'delete') {
      throw ArgumentError.value(
        operation,
        'write.op',
        'must be upsert or delete',
      );
    }
    if (baseVersion is! int || baseVersion < 0) {
      throw ArgumentError.value(
        baseVersion,
        'write.base_version',
        'must be non-negative',
      );
    }
    final suppliedKey = write['key'];
    if (suppliedKey != null &&
        (suppliedKey is! String || suppliedKey.trim().isEmpty)) {
      throw ArgumentError.value(
        suppliedKey,
        'write.key',
        'must be a nonempty string',
      );
    }
    final key = suppliedKey is String
        ? suppliedKey
        : '$table:$id:$operation:$baseVersion';
    if (options?.idempotencyKey != null && options!.idempotencyKey != key) {
      throw ArgumentError(
        'explicit idempotency key must match the sync write key',
      );
    }
    final prefix = pathPrefix.replaceAll(RegExp(r'/+$'), '');
    final wireWrite = <String, Object?>{
      'id': id,
      'table': table,
      'op': operation,
      'payload': write['payload'],
      'base_version': baseVersion,
      'key': key,
    };
    final response = await _request(
      'POST',
      '$prefix/${_enc(table)}',
      wireWrite,
      _withIdempotencyKey(options, key),
    );
    if (response is! Map) {
      throw const FormatException('sync acknowledgement must be an object');
    }
    final acknowledgement = Map<String, Object?>.from(response);
    final committedVersion = acknowledgement['committed_version'];
    if (acknowledgement['id'] != id) {
      throw const FormatException(
        'sync acknowledgement id does not match the queued write',
      );
    }
    if (committedVersion is! int || committedVersion < 0) {
      throw const FormatException(
        'sync acknowledgement has an invalid committed_version',
      );
    }
    return acknowledgement;
  }

  FiduciaSyncJsonSender syncSender({
    String pathPrefix = '/api/customer/sync',
    FiduciaRequestOptions? options,
  }) =>
      (write) => syncWrite(write, pathPrefix: pathPrefix, options: options);

  Future<Map<String, Object?>> syncPull(
    String table,
    int cursor, {
    String pathPrefix = '/api/customer/sync',
    int limit = 500,
    FiduciaRequestOptions? options,
  }) async {
    if (table.trim().isEmpty) {
      throw ArgumentError.value(table, 'table', 'must be nonempty');
    }
    if (cursor < 0 || limit < 1 || limit > 1000) {
      throw ArgumentError('cursor or sync pull limit is outside its range');
    }
    final prefix = pathPrefix.replaceAll(RegExp(r'/+$'), '');
    final response = await _request(
      'GET',
      '$prefix/${_enc(table)}?cursor=$cursor&limit=$limit',
      null,
      options,
    );
    if (response is! Map || response['changes'] is! List) {
      throw const FormatException('sync pull response has no changes array');
    }
    final nextCursor = response['next_cursor'];
    final hasMore = response['has_more'];
    if (nextCursor is! int || nextCursor < cursor || hasMore is! bool) {
      throw const FormatException('sync pull response has an invalid cursor');
    }
    final changes = (response['changes'] as List).map((value) {
      if (value is! Map) {
        throw const FormatException('sync change must be an object');
      }
      final change = Map<String, Object?>.from(value);
      change['at_ms'] = change['at_ms'] is int ? change['at_ms'] : 0;
      if (change['sync_sequence'] is! int && change['sequence'] is int) {
        change['sync_sequence'] = change['sequence'];
      }
      change.remove('sequence');
      return change;
    }).toList(growable: false);
    return {'changes': changes, 'next_cursor': nextCursor, 'has_more': hasMore};
  }

  FiduciaSyncJsonPull syncPuller(
    String table, {
    String pathPrefix = '/api/customer/sync',
    FiduciaRequestOptions? options,
  }) =>
      (cursor, limit) => syncPull(
            table,
            cursor,
            pathPrefix: pathPrefix,
            limit: limit,
            options: options,
          );

  // --- locks & semaphores ---
  Future<dynamic> lockAcquire(
    String key, {
    int? ttlMs,
    bool wait = true,
    int max = 1,
    FiduciaRequestOptions? options,
  }) =>
      _lockAcquireWithWait(
        key,
        ttlMs: ttlMs,
        wait: wait,
        max: max,
        options: options,
      );
  Future<dynamic> tryLock(
    String key, {
    int? ttlMs,
    int max = 1,
    FiduciaRequestOptions? options,
  }) =>
      _lockAcquireWithWait(
        key,
        ttlMs: ttlMs,
        wait: false,
        max: max,
        options: options,
      );
  Future<dynamic> mustLock(
    String key, {
    int? ttlMs,
    int max = 1,
    FiduciaRequestOptions? options,
  }) =>
      _lockAcquireWithWait(
        key,
        ttlMs: ttlMs,
        wait: true,
        max: max,
        options: options,
      );
  Future<dynamic> lock(
    String key, {
    int? ttlMs,
    int max = 1,
    FiduciaRequestOptions? options,
  }) =>
      mustLock(key, ttlMs: ttlMs, max: max, options: options);
  Future<dynamic> _lockAcquireWithWait(
    String key, {
    int? ttlMs,
    required bool wait,
    required int max,
    FiduciaRequestOptions? options,
  }) =>
      _request(
        'POST',
        '/v1/locks/acquire',
        {'key': key, 'ttl_ms': ttlMs, 'wait': wait, 'max': max},
        options,
        true,
      );
  Future<dynamic> lockRelease(String key, String holder, int fencingToken) =>
      _request('POST', '/v1/locks/release', {
        'holder': holder,
        'fencing_token': fencingToken,
      });
  Future<dynamic> semaphoreAcquire(
    String key, {
    int? ttlMs,
    bool wait = true,
    int max = 2,
    FiduciaRequestOptions? options,
  }) =>
      _semaphoreAcquireWithWait(
        key,
        ttlMs: ttlMs,
        wait: wait,
        max: max,
        options: options,
      );
  Future<dynamic> trySemaphore(
    String key, {
    int? ttlMs,
    int max = 2,
    FiduciaRequestOptions? options,
  }) =>
      _semaphoreAcquireWithWait(
        key,
        ttlMs: ttlMs,
        wait: false,
        max: max,
        options: options,
      );
  Future<dynamic> mustSemaphore(
    String key, {
    int? ttlMs,
    int max = 2,
    FiduciaRequestOptions? options,
  }) =>
      _semaphoreAcquireWithWait(
        key,
        ttlMs: ttlMs,
        wait: true,
        max: max,
        options: options,
      );
  Future<dynamic> semaphore(
    String key, {
    int? ttlMs,
    int max = 2,
    FiduciaRequestOptions? options,
  }) =>
      mustSemaphore(key, ttlMs: ttlMs, max: max, options: options);
  Future<dynamic> _semaphoreAcquireWithWait(
    String key, {
    int? ttlMs,
    required bool wait,
    required int max,
    FiduciaRequestOptions? options,
  }) =>
      _request(
        'POST',
        '/v1/semaphores/acquire',
        {'key': key, 'ttl_ms': ttlMs, 'wait': wait, 'limit': max < 2 ? 2 : max},
        options,
        true,
      );
  Future<dynamic> semaphoreRelease(
    String key,
    String holder,
    int fencingToken,
  ) =>
      _request('POST', '/v1/semaphores/release', {
        'key': key,
        'holder': holder,
        'fencing_token': fencingToken,
      });

  // --- reader-writer locks ---
  Future<dynamic> rwAcquireRead(String key, {int? ttlMs, bool wait = true}) =>
      _request('POST', '/v1/rw/${_enc(key)}/read', {
        'ttl_ms': ttlMs,
        'wait': wait,
      });
  Future<dynamic> rwEndRead(String key, String lockId) =>
      _request('POST', '/v1/rw/${_enc(key)}/read/end', {'lock_id': lockId});
  Future<dynamic> rwAcquireWrite(String key, {int? ttlMs, bool wait = true}) =>
      _request('POST', '/v1/rw/${_enc(key)}/write', {
        'ttl_ms': ttlMs,
        'wait': wait,
      });
  Future<dynamic> rwEndWrite(String key, String lockId) =>
      _request('POST', '/v1/rw/${_enc(key)}/write/end', {'lock_id': lockId});

  // --- config KV ---
  Future<dynamic> kvGet(String key) =>
      _request('GET', '/v1/kv?key=${_enc(key)}');
  Future<dynamic> kvPut(String key, String value, {int? ttlMs}) => _request(
        'PUT',
        '/v1/kv?key=${_enc(key)}',
        {'value': value, 'ttl_ms': ttlMs},
      );
  Future<dynamic> kvDelete(String key) =>
      _request('DELETE', '/v1/kv?key=${_enc(key)}');
  Future<dynamic> kvList(String prefix) =>
      _request('GET', '/v1/kv?prefix=${_enc(prefix)}');

  // --- leader election ---
  Future<dynamic> electionCampaign(String name, String candidate, int ttlMs) =>
      _request('POST', '/v1/elections/${_enc(name)}/campaign', {
        'candidate': candidate,
        'ttl_ms': ttlMs,
      });
  Future<dynamic> electionRenew(
    String name,
    String candidate,
    int fencingToken,
  ) =>
      _request('POST', '/v1/elections/${_enc(name)}/renew', {
        'candidate': candidate,
        'fencing_token': fencingToken,
      });
  Future<dynamic> electionResign(
    String name,
    String candidate,
    int fencingToken,
  ) =>
      _request('POST', '/v1/elections/${_enc(name)}/resign', {
        'candidate': candidate,
        'fencing_token': fencingToken,
      });
  Future<dynamic> electionGet(String name) =>
      _request('GET', '/v1/elections/${_enc(name)}');

  // --- service discovery ---
  Future<dynamic> serviceRegister(
    String service,
    String instanceId,
    String address,
    int ttlMs,
  ) =>
      _request(
        'PUT',
        '/v1/services/${_enc(service)}/instances/${_enc(instanceId)}',
        {'address': address, 'ttl_ms': ttlMs},
      );
  Future<dynamic> serviceHeartbeat(String service, String instanceId) =>
      _request(
        'POST',
        '/v1/services/${_enc(service)}/instances/${_enc(instanceId)}/heartbeat',
      );
  Future<dynamic> serviceDeregister(String service, String instanceId) =>
      _request(
        'DELETE',
        '/v1/services/${_enc(service)}/instances/${_enc(instanceId)}',
      );
  Future<dynamic> serviceInstances(String service) =>
      _request('GET', '/v1/services/${_enc(service)}');
  Future<dynamic> serviceList() => _request('GET', '/v1/services');
}

File: clients/matlab/Fiducia.m
% Fiducia HTTP client (MATLAB). Zero-dependency -- matlab.net.http + jsonencode/jsondecode (built-in).
% Implements PROTOCOL.md.
%
%   c = Fiducia('https://api.fiducia.cloud');
%   lk = c.lockAcquire('orders/checkout', 'ttl_ms', 30000);
%   c.lockRelease('orders/checkout', 'worker-a', lk.result.output.fencing_token);
%
% Every method returns the decoded JSON response (a struct/cell/array via
% jsondecode; [] for an empty body). On HTTP status >= 300 it raises an
% MException with identifier 'fiducia:httpError' whose message carries the
% numeric status and the raw response body. Requires MATLAB R2019b or newer
% (arguments blocks). Name-value options use MATLAB's name-value syntax, e.g.
% c.lockAcquire('k', 'ttl_ms', 30000) or, on R2021a+, c.lockAcquire('k', ttl_ms=30000).

classdef Fiducia < handle
    % Fiducia  Thin HTTP client for the fiducia.cloud coordination API.

    properties
        BaseUrl            % Base URL, trailing slash trimmed.
        RequestTimeout = 30  % Request timeout in seconds (connect + response). Set [] to use the matlab.net.http default.
    end

    methods
        function obj = Fiducia(baseUrl)
            arguments
                baseUrl (1,:) char
            end
            obj.BaseUrl = regexprep(baseUrl, '/+$', '');
        end

        % --- misc ---
        function out = health(obj)
            out = obj.doRequest('GET', '/healthz');
        end

        function out = status(obj)
            out = obj.doRequest('GET', '/v1/status');
        end

        % --- locks ---
        function out = lockGet(obj, key)
            out = obj.doRequest('GET', ['/v1/locks?key=', obj.enc(key)]);
        end

        function out = lockAcquire(obj, key, opts)
            arguments
                obj
                key
                opts.holder = []
                opts.ttl_ms = []
                opts.wait (1,1) logical = true
            end
            body = struct('key', key);
            body = obj.addOpt(body, 'holder', opts.holder);
            body = obj.addOpt(body, 'ttl_ms', opts.ttl_ms);
            body.wait = opts.wait;
            out = obj.doRequest('POST', '/v1/locks/acquire', body);
        end

        function out = lockAcquireMany(obj, keys, opts)
            arguments
                obj
                keys
                opts.holder = []
                opts.ttl_ms = []
                opts.wait (1,1) logical = true
            end
            body = struct();
            body.keys = cellstr(keys);  % union lock over a string array
            body = obj.addOpt(body, 'holder', opts.holder);
            body = obj.addOpt(body, 'ttl_ms', opts.ttl_ms);
            body.wait = opts.wait;
            out = obj.doRequest('POST', '/v1/locks/acquire', body);
        end

        function out = tryLock(obj, key, opts)
            arguments
                obj
                key
                opts.holder = []
                opts.ttl_ms = []
            end
            out = obj.lockAcquire(key, 'holder', opts.holder, 'ttl_ms', opts.ttl_ms, 'wait', false);
        end

        function grant = mustLock(obj, key, opts)
            % mustLock  Blocking acquire. The server does NOT hold the connection
            % on wait=true (it reserves a FIFO slot and returns immediately), so
            % this acquires then POLLS lockGet until this holder holds the lock,
            % or raises 'fiducia:lockTimeout' after max_wait_ms. Returns a held
            % grant struct (key, holder, fencing_token, lease_expires_ms) that you
            % can pass straight to lockRelease.
            arguments
                obj
                key
                opts.holder = []
                opts.ttl_ms = []
                opts.max_wait_ms (1,1) double = 30000
                opts.retry_interval_ms (1,1) double = 250
                opts.max_retries = []
            end
            grant = obj.acquireLockBlocking(key, opts.holder, opts.ttl_ms, ...
                opts.max_wait_ms, opts.retry_interval_ms, opts.max_retries);
        end

        function grant = lock(obj, key, opts)
            % lock  Alias for mustLock (blocking acquire that polls until held).
            arguments
                obj
                key
                opts.holder = []
                opts.ttl_ms = []
                opts.max_wait_ms (1,1) double = 30000
                opts.retry_interval_ms (1,1) double = 250
                opts.max_retries = []
            end
            grant = obj.mustLock(key, 'holder', opts.holder, 'ttl_ms', opts.ttl_ms, ...
                'max_wait_ms', opts.max_wait_ms, 'retry_interval_ms', opts.retry_interval_ms, ...
                'max_retries', opts.max_retries);
        end

        function out = lockRelease(obj, key, holder, fencing_token)
            % key is accepted for symmetry but is NOT sent in the body.
            body = struct('holder', holder, 'fencing_token', fencing_token);
            out = obj.doRequest('POST', '/v1/locks/release', body);
        end

        % --- semaphores ---
        function out = semaphoreGet(obj, key)
            out = obj.doRequest('GET', ['/v1/semaphores?key=', obj.enc(key)]);
        end

        function out = semaphoreAcquire(obj, key, limit, opts)
            arguments
                obj
                key
                limit
                opts.holder = []
                opts.ttl_ms = []
                opts.wait (1,1) logical = true
            end
            body = struct('key', key);
            body = obj.addOpt(body, 'holder', opts.holder);
            body = obj.addOpt(body, 'ttl_ms', opts.ttl_ms);
            body.limit = limit;
            body.wait = opts.wait;
            out = obj.doRequest('POST', '/v1/semaphores/acquire', body);
        end

        function out = trySemaphore(obj, key, limit, opts)
            arguments
                obj
                key
                limit
                opts.holder = []
                opts.ttl_ms = []
            end
            out = obj.semaphoreAcquire(key, limit, 'holder', opts.holder, 'ttl_ms', opts.ttl_ms, 'wait', false);
        end

        function out = mustSemaphore(obj, key, limit, opts)
            arguments
                obj
                key
                limit
                opts.holder = []
                opts.ttl_ms = []
            end
            out = obj.semaphoreAcquire(key, limit, 'holder', opts.holder, 'ttl_ms', opts.ttl_ms, 'wait', true);
        end

        function out = semaphore(obj, key, limit, opts)
            % semaphore  Alias for mustSemaphore (blocking acquire, wait=true).
            arguments
                obj
                key
                limit
                opts.holder = []
                opts.ttl_ms = []
            end
            out = obj.mustSemaphore(key, limit, 'holder', opts.holder, 'ttl_ms', opts.ttl_ms);
        end

        function out = semaphoreRelease(obj, key, holder, fencing_token)
            body = struct('key', key, 'holder', holder, 'fencing_token', fencing_token);
            out = obj.doRequest('POST', '/v1/semaphores/release', body);
        end

        % --- idempotency ---
        function out = idempotencyGet(obj, key)
            out = obj.doRequest('GET', ['/v1/idempotency?key=', obj.enc(key)]);
        end

        function out = idempotencyClaim(obj, key, opts)
            arguments
                obj
                key
                opts.owner = []
                opts.ttl_ms = []
                opts.ttl = []
                opts.metadata = []  % arbitrary JSON object (struct)
            end
            body = struct('key', key);
            body = obj.addOpt(body, 'owner', opts.owner);
            body = obj.addOpt(body, 'ttl_ms', opts.ttl_ms);
            body = obj.addOpt(body, 'ttl', opts.ttl);
            body = obj.addOpt(body, 'metadata', opts.metadata);
            out = obj.doRequest('POST', '/v1/idempotency/claim', body);
        end

        function out = idempotencyComplete(obj, key, owner, fencing_token, opts)
            arguments
                obj
                key
                owner
                fencing_token
                opts.result = []  % arbitrary JSON object (struct)
            end
            body = struct('key', key, 'owner', owner, 'fencing_token', fencing_token);
            body = obj.addOpt(body, 'result', opts.result);
            out = obj.doRequest('POST', '/v1/idempotency/complete', body);
        end

        % --- reader-writer locks ---
        function out = rwAcquireRead(obj, key, opts)
            arguments
                obj
                key
                opts.ttl_ms = []
                opts.wait (1,1) logical = true
            end
            body = struct();
            body = obj.addOpt(body, 'ttl_ms', opts.ttl_ms);
            body.wait = opts.wait;
            out = obj.doRequest('POST', ['/v1/rw/', obj.enc(key), '/read'], body);
        end

        function out = rwEndRead(obj, key, lock_id)
            body = struct('lock_id', lock_id);
            out = obj.doRequest('POST', ['/v1/rw/', obj.enc(key), '/read/end'], body);
        end

        function out = rwAcquireWrite(obj, key, opts)
            arguments
                obj
                key
                opts.ttl_ms = []
                opts.wait (1,1) logical = true
            end
            body = struct();
            body = obj.addOpt(body, 'ttl_ms', opts.ttl_ms);
            body.wait = opts.wait;
            out = obj.doRequest('POST', ['/v1/rw/', obj.enc(key), '/write'], body);
        end

        function out = rwEndWrite(obj, key, lock_id)
            body = struct('lock_id', lock_id);
            out = obj.doRequest('POST', ['/v1/rw/', obj.enc(key), '/write/end'], body);
        end

        % --- config KV ---
        function out = kvGet(obj, key)
            out = obj.doRequest('GET', ['/v1/kv?key=', obj.enc(key)]);
        end

        function out = kvPut(obj, key, value, opts)
            arguments
                obj
                key
                value
                opts.ttl_ms = []
                opts.prev_revision = []
            end
            body = struct();
            body.value = value;
            body = obj.addOpt(body, 'ttl_ms', opts.ttl_ms);
            body = obj.addOpt(body, 'prev_revision', opts.prev_revision);
            out = obj.doRequest('PUT', ['/v1/kv?key=', obj.enc(key)], body);
        end

        function out = kvDelete(obj, key)
            out = obj.doRequest('DELETE', ['/v1/kv?key=', obj.enc(key)]);
        end

        function out = kvList(obj, prefix)
            out = obj.doRequest('GET', ['/v1/kv?prefix=', obj.enc(prefix)]);
        end

        % --- rate limiting ---
        function out = rateLimitGet(obj, tenant, key)
            out = obj.doRequest('GET', ['/v1/rate-limit/', obj.enc(tenant), '/', obj.enc(key)]);
        end

        function out = rateLimitCheck(obj, tenant, key, algorithm, limit, window_ms, opts)
            arguments
                obj
                tenant
                key
                algorithm
                limit
                window_ms
                opts.refill_per_second = []
                opts.cost = []
            end
            body = struct('algorithm', algorithm, 'limit', limit, 'window_ms', window_ms);
            body = obj.addOpt(body, 'refill_per_second', opts.refill_per_second);
            body = obj.addOpt(body, 'cost', opts.cost);
            out = obj.doRequest('POST', ['/v1/rate-limit/', obj.enc(tenant), '/', obj.enc(key), '/check'], body);
        end

        % --- cron & scheduling ---
        function out = scheduleGet(obj, name)
            out = obj.doRequest('GET', ['/v1/cron/schedules/', obj.enc(name)]);
        end

        function out = scheduleUpsert(obj, name, target, opts)
            arguments
                obj
                name
                target  % arbitrary JSON object (struct), e.g. struct('kind','webhook','url','...')
                opts.cron = []
                opts.one_shot_at_ms = []
                opts.delivery = []
                opts.max_retries = []
            end
            body = struct();
            body.target = target;
            body = obj.addOpt(body, 'cron', opts.cron);
            body = obj.addOpt(body, 'one_shot_at_ms', opts.one_shot_at_ms);
            body = obj.addOpt(body, 'delivery', opts.delivery);
            body = obj.addOpt(body, 'max_retries', opts.max_retries);
            out = obj.doRequest('PUT', ['/v1/cron/schedules/', obj.enc(name)], body);
        end

        function out = scheduleRecordRun(obj, name, fire_id, opts)
            arguments
                obj
                name
                fire_id
                opts.fired_at_ms = []
            end
            body = struct('fire_id', fire_id);
            body = obj.addOpt(body, 'fired_at_ms', opts.fired_at_ms);
            out = obj.doRequest('POST', ['/v1/cron/schedules/', obj.enc(name), '/runs'], body);
        end

        function out = scheduleHistory(obj, name)
            out = obj.doRequest('GET', ['/v1/cron/schedules/', obj.enc(name), '/history']);
        end

        % --- leader election ---
        function out = electionGet(obj, name)
            out = obj.doRequest('GET', ['/v1/elections/', obj.enc(name)]);
        end

        function out = electionCampaign(obj, name, candidate, ttl_ms, opts)
            arguments
                obj
                name
                candidate
                ttl_ms
                opts.metadata = []
            end
            body = struct('candidate', candidate, 'ttl_ms', ttl_ms);
            body = obj.addOpt(body, 'metadata', opts.metadata);
            out = obj.doRequest('POST', ['/v1/elections/', obj.enc(name), '/campaign'], body);
        end

        function out = electionRenew(obj, name, candidate, fencing_token)
            body = struct('candidate', candidate, 'fencing_token', fencing_token);
            out = obj.doRequest('POST', ['/v1/elections/', obj.enc(name), '/renew'], body);
        end

        function out = electionResign(obj, name, candidate, fencing_token)
            body = struct('candidate', candidate, 'fencing_token', fencing_token);
            out = obj.doRequest('POST', ['/v1/elections/', obj.enc(name), '/resign'], body);
        end

        % --- service discovery ---
        function out = serviceInstances(obj, service)
            out = obj.doRequest('GET', ['/v1/services/', obj.enc(service)]);
        end

        function out = serviceRegister(obj, service, instance_id, address, ttl_ms, opts)
            arguments
                obj
                service
                instance_id
                address
                ttl_ms
                opts.metadata = []
            end
            body = struct('address', address, 'ttl_ms', ttl_ms);
            body = obj.addOpt(body, 'metadata', opts.metadata);
            out = obj.doRequest('PUT', ['/v1/services/', obj.enc(service), '/instances/', obj.enc(instance_id)], body);
        end

        function out = serviceHeartbeat(obj, service, instance_id, opts)
            arguments
                obj
                service
                instance_id
                opts.ttl_ms = []
            end
            body = struct();
            body = obj.addOpt(body, 'ttl_ms', opts.ttl_ms);
            out = obj.doRequest('POST', ['/v1/services/', obj.enc(service), '/instances/', obj.enc(instance_id), '/heartbeat'], body);
        end

        function out = serviceDeregister(obj, service, instance_id)
            out = obj.doRequest('DELETE', ['/v1/services/', obj.enc(service), '/instances/', obj.enc(instance_id)]);
        end

        function out = serviceList(obj)
            out = obj.doRequest('GET', '/v1/services');
        end
    end

    methods (Access = private)
        function out = doRequest(obj, method, path, body)
            % Core request. Pass a 4th argument to send a JSON body.
            url = [obj.BaseUrl, path];
            if nargin >= 4
                header = matlab.net.http.HeaderField('Content-Type', 'application/json');
                % Encode with jsonencode and hand matlab.net.http raw uint8 bytes so
                % it transmits them verbatim (a char/string body would be re-encoded).
                payloadBytes = unicode2native(jsonencode(body), 'UTF-8');
                request = matlab.net.http.RequestMessage(method, header, ...
                    matlab.net.http.MessageBody(payloadBytes));
            else
                request = matlab.net.http.RequestMessage(method);
            end

            % MaxRedirects=0: do NOT auto-follow 3xx. Following a redirect on a
            % mutating POST/PUT/DELETE could re-submit and duplicate the operation
            % (e.g. a lock grant / FIFO queue slot). A 3xx is >= 300, so it surfaces
            % through the normal error path below rather than being followed.
            if isempty(obj.RequestTimeout)
                options = matlab.net.http.HTTPOptions('SavePayload', true, ...
                    'MaxRedirects', 0);
            else
                options = matlab.net.http.HTTPOptions('SavePayload', true, ...
                    'MaxRedirects', 0, ...
                    'ConnectTimeout', obj.RequestTimeout, ...
                    'ResponseTimeout', obj.RequestTimeout);
            end

            % send() returns the response for any HTTP status (it does not throw
            % on 4xx/5xx), so the status code is read cleanly below.
            response = request.send(matlab.net.URI(url), options);
            statusCode = double(response.StatusCode);
            rawBody = obj.rawText(response);
            if statusCode >= 300
                obj.raiseHttpError(statusCode, rawBody);
            end
            out = obj.decodeText(rawBody);
        end

        function s = rawText(~, response)
            % Return the response body as a char row (raw, undecoded).
            payload = response.Body.Payload;
            if ~isempty(payload)
                s = native2unicode(reshape(payload, 1, []), 'UTF-8');
                return;
            end
            data = response.Body.Data;
            if isempty(data)
                s = '';
            elseif ischar(data)
                s = data;
            elseif isstring(data)
                s = char(data);
            else
                % matlab.net.http auto-decoded a JSON body; re-serialize so the
                % caller path can jsondecode uniformly.
                s = jsonencode(data);
            end
        end

        function data = decodeText(~, s)
            t = strtrim(s);
            if isempty(t)
                data = [];
            else
                data = jsondecode(t);
            end
        end

        function raiseHttpError(~, statusCode, rawBody)
            if isempty(rawBody)
                message = sprintf('fiducia: HTTP %d', statusCode);
            else
                message = sprintf('fiducia: HTTP %d: %s', statusCode, rawBody);
            end
            error(MException('fiducia:httpError', '%s', message));
        end

        function b = addOpt(~, b, name, value)
            % Add an optional field to the body only when a value was supplied
            % (omit-nulls; matters for CAS semantics).
            if ~isempty(value)
                b.(name) = value;
            end
        end

        function e = enc(~, s)
            % Percent-encode a string for a URL path segment or query value.
            % Encodes every octet except the RFC 3986 unreserved set.
            s = char(string(s));
            if isempty(s)
                e = '';
                return;
            end
            bytes = unicode2native(s, 'UTF-8');
            unreserved = uint8(['A':'Z', 'a':'z', '0':'9', '-', '_', '.', '~']);
            parts = cell(1, numel(bytes));
            for i = 1:numel(bytes)
                b = bytes(i);
                if any(b == unreserved)
                    parts{i} = char(b);
                else
                    parts{i} = sprintf('%%%02X', b);
                end
            end
            e = strjoin(parts, '');
        end
    end
end
